### PR TITLE
refactor gradient calculations to use `DerivativeSurfaceMesh`

### LINKED ISSUE #1111: Generalize VJP calculation
Potentially useful to speed things up, will prototype one possible approach:
* Get coordinates for all integration points in the structure. 
* Downsample E fwd and adj to the integration coordinates.
* Continue as normal, selecting instead of interpolating.

### LINKED ISSUE #1117: Adjoint `PolySlab.slab_bounds` support
Add the ability to differentiate with respect to the `PolySlab.slab_bounds`.

Will involve computing the VJP for this field using the adjoint shifting boundary formalism. 

Then this VJP value needs to be returned properly in the `JaxPolySlab.store_vjp()` method.

See some hints in the medium VJP functions for how to handle arbitrary shape evaluation.

Also, add a test to the backend adjoint tests to confirm the gradient accuracy.

### LINKED ISSUE #1985: autograd support for `PolySlab.slab_bounds`
Would be a good idea to include this, in case users want to compute sensitivity with respect to polyslab bounds.

Should be pretty straightforward (famous last words :P)

### PR DESCRIPTION
**Pull request by tylerflex**
Part 1 of an effort to generalize the VJP calculations so the Geometry and Medium just need to construct a single `DerivativeSurfaceMesh` for every field. The `DerivativeInfo` then has a method that computes derivatives given a `DerivativeSurfaceMesh`.

### CODE REVIEW

**Comment by tylerflex** [comment]
Let's also get this one in. Looks like I have a good amount of merge conflicts to sort out first, but I'll work on that today and re-test

**Comment by tylerflex** [comment]
did a bit of testing of this and it seems like it still works. going to leave it unsquashed for now but should be ready to go after reviews @momchil-flex @yaugenst-flex

**Comment by tylerflex** [comment]
I just noticed an issue with the epsilon grabbing from the other PR after the rebase, so don't merge this yet as I'll fix it tomorrow morning

**Comment by tylerflex** [comment]
alright fixed it, turned out when there were `inf` in the `PolySlab.slab_bounds` or `Cylinder.length`, some weird things would happen after the rebase. These are handled properly now.

File: tidy3d/components/autograd/derivative_utils.py
Comment (yaugenst-flex) [inline tidy3d/components/autograd/derivative_utils.py]: Additional quotes slipped int?

File: tidy3d/components/autograd/derivative_utils.py
Comment (yaugenst-flex) [inline tidy3d/components/autograd/derivative_utils.py]: Feels like `edge_index` is better than `i`?

File: tidy3d/components/autograd/derivative_utils.py
Comment (yaugenst-flex) [inline tidy3d/components/autograd/derivative_utils.py]: I think this still needs `bounds_error=False`, because according to [xarray's docs](https://docs.xarray.dev/en/stable/generated/xarray.DataArray.interp.html), `da.interp()` will choose the default value from either [interp1d](https://docs.scipy.org/doc/scipy/reference/generated/scipy.interpolate.interp1d.html) or [interpn](https://docs.scipy.org/doc/scipy/reference/generated/scipy.interpolate.interpn.html), the latter has a default of `bounds_error=True`, so this will fail for >1d interpolation.

File: tidy3d/components/data/sim_data.py
Comment (yaugenst-flex) [inline tidy3d/components/data/sim_data.py]: Unrelated to this PR, but this is not in `pre/2.8`, we should probably do another merge from develop? We probably shouldn't be doing that merge together with this PR.

File: tidy3d/components/geometry/base.py
Comment (yaugenst-flex) [inline tidy3d/components/geometry/base.py]: Why is this necessary? Shouldn't the code below still work either way?

**Review by yaugenst-flex — COMMENTED** [review]
Really nice to have this 🎉
Only some minor comments. Biggest issue for me is, this is basically a merge from develop right? I think we should do that separately, there are a lot of unrelated changes in here.

**Comment by momchil-flex** [comment]
Yeah @tylerflex I think you need to drop the first 6 commits (then maybe clean up the rest?)

File: tidy3d/components/autograd/derivative_utils.py
Comment (tylerflex) [reply tidy3d/components/autograd/derivative_utils.py]: ah yea, actually this snuck in because I was testing whether xarray was freaking out with a dim with more than one character, will revert.

File: tidy3d/components/geometry/base.py
Comment (tylerflex) [reply tidy3d/components/geometry/base.py]: this might have been a relic from when I was testing some validation errors when slab_bounds were traced. maybe we won't need them now.

File: tidy3d/components/autograd/derivative_utils.py
Comment (tylerflex) [reply tidy3d/components/autograd/derivative_utils.py]: hm, but the thing is that I'm doing > 1d interpolation here,  and often out of bounds, so why wouldn't it be failing?

File: tidy3d/components/autograd/derivative_utils.py
Comment (tylerflex) [reply tidy3d/components/autograd/derivative_utils.py]: fixed

File: tidy3d/components/autograd/derivative_utils.py
Comment (tylerflex) [reply tidy3d/components/autograd/derivative_utils.py]: well it worked with `bounds_error=False` so I'll just put it in for extra security

**Comment by tylerflex** [comment]
ok I cleaned up the commits into just 2, addressed Yannick's comments, but now ruff is complaining (but not locally, so that's a bit weird)

File: tidy3d/components/autograd/derivative_utils.py
Comment (yaugenst) [reply tidy3d/components/autograd/derivative_utils.py]: Ok interesting, wouldn't have expected that. But still yeah I think better to keep it explicit.

**Review by yaugenst-flex — APPROVED** [review]

File: tidy3d/components/autograd/derivative_utils.py
Comment (momchil-flex) [inline tidy3d/components/autograd/derivative_utils.py:30]: Maybe move the last paragraph here to further clarify this sign/direction point? I first wrote a comment about it and then saw you've answered it in the last paragraph.

File: tidy3d/components/autograd/derivative_utils.py
Comment (momchil-flex) [inline tidy3d/components/autograd/derivative_utils.py]: Capitalize `user`?

File: tidy3d/components/structure.py
Comment (momchil-flex) [inline tidy3d/components/structure.py:251]: Whenever changes are made to the frontend autograd components we need to be careful about the backend too. Seems like [this](https://github.com/flexcompute/tidy3d-core/blob/000aab3e18c9648fcee4ef93195bb76aaaed5a48/tidy3d_backend/metadata.py#L650) needs to be updated and I'm not sure how?

File: tidy3d/components/geometry/polyslab.py
Comment (momchil-flex) [inline tidy3d/components/geometry/polyslab.py]: So at some point we purposefully changed some validators to use `isclose` instead of `np.isclose` because the latter creates a numpy array which can have significant overhead if you e.g. have a metalens with a million polyslabs. We have [this test](https://github.com/flexcompute/tidy3d-core/blob/000aab3e18c9648fcee4ef93195bb76aaaed5a48/tests/test_simulation.py#L8-L38) on the backend about this (it could easily be moved to the frontend if we wanted). I tried a few times using `pre/2.8` and it would run in about 6.5-7s. With your branch, it runs in about 8.5-9s, which is potentially acceptable but still somewhat significant. Increasing the number of geometries seems to have a linear scaling so it's always about 30% slower.

This consideration was definitely much more important in some of the mesher code, where the scaling is with number of structures squared. But yeah even here it has an effect.

Is it necessary to change to numpy?

File: tidy3d/components/autograd/derivative_utils.py
Comment (tylerflex) [reply tidy3d/components/autograd/derivative_utils.py:30]: sure, sounds good. I think I also ended up writing that last paragraph because after re-reading I realized it needed more clarification, so makes sense to put there.

File: tidy3d/components/autograd/derivative_utils.py
Comment (tylerflex) [reply tidy3d/components/autograd/derivative_utils.py]: done

File: tidy3d/components/structure.py
Comment (tylerflex) [reply tidy3d/components/structure.py:251]: ah, right. I think the fix is to pass the `sim_fields_keys` into the `make_adjoint_monitors`. Since you have the optimizer, that would mean getting

```py
sim = optimizer.design.simulation
traced_fields_sim = setup_run(simulation=sim)
sim_fields_keys = list(traced_fields_sim.keys())
```

Comment (momchil-flex) [reply tidy3d/components/structure.py:251]: I don't think `setup_run` will work if this is not called inside a `grad` call?

File: tidy3d/components/geometry/polyslab.py
Comment (tylerflex) [reply tidy3d/components/geometry/polyslab.py]: I changed them all back and things are passing, so fine by me. I think originally I had changed them because I was running into validation errors with `slab_bounds` and it was really hard to pinpoint what was failing. and so I ended up basically changing a ton of things and removing validators before I found it. this must have slipped into the PR.

File: tidy3d/components/structure.py
Comment (tylerflex) [reply tidy3d/components/structure.py:251]: not sure, guess you could just do

```

    return simulation.strip_traced_fields(
        include_untraced_data_arrays=False, starting_path=("structures",)
    )

```
which is what setup run does.

Comment (momchil-flex) [reply tidy3d/components/structure.py:251]: But the point is that there are no traced fields since it's not a simulation called in `grad` no?

Comment (momchil-flex) [reply tidy3d/components/structure.py:251]: Basically what we're trying to do there is get the gradient monitors from the static definition of the InverseDesign object.

Comment (tylerflex) [reply tidy3d/components/structure.py:251]: ah ok. yea I guess we dont know. but we can effectively "fake" it then, because we know it's a Box with a CustomMedium and we need the `permittivity` data.

So I think one approach could just be to pass
```
field_keys=(f'structures/{(len(sim.structures) - 1}/medium/permittivity', )
```
into the `make_adjoint_monitors` call. Or you can just make the monitors manually since again we know it's just a box so it needs a similarly sized field and permittivity monitor

Comment (momchil-flex) [reply tidy3d/components/structure.py:251]: Using the first approach, works for now, obviously should be dealt with better in the future..

**Pull request merged**
